doc/stratagus.6
.TH STRATAGUS 6 "Oct 2011" "Stratagus"
.SH NAME
stratagus \- Strategy Gaming Engine
.SH SYNOPSIS
.B stratagus
.I [-a] [-c file.lua] [-d datapath] [-D depth] [-e] [-E file.lua] [-F|-W] [-G options] [-h] [-I addr] [-l]
.I [-N name] [-o|-O] [-p] [-P port] [-s sleep] [-S speed] [-v mode] [-x scaler-idx] [-Z] [map.smp|map.smp.gz]
.SH "DESCRIPTION"
This manual page documents briefly the flags that you can give to
.BR stratagus
, formerly known as FreeCraft, formerly known as ALE Clone.
.PP
.B stratagus
is a powerful game engine for building realtime strategy games, a la AoE or Warcraft.
.SH OPTIONS
A summary of options is included below.
.TP
.B \-a
Enables asserts check in engine code (for debugging)
.TP
.B \-c file.lua
Choose a different stratagus start config file.
.TP
.B \-d datapath
Choose a different directory to use for game data files.
.TP
.B \-D depth
Set the video mode depth, often called 'bpp' or 'bits per pixel'. This
option is only valid for Win32/NT.
.TP
.B \-e
Start the editor included in Stratagus.
.TP
.B \-E
Choose a different editor config file.
.TP
.B \-F
With this option you can start Stratagus in full screen video mode.
.TP
.B \-G "lua options"
The argument is passed as argument string to the lua game script. Take care to
quote if spaces are used. This can be used in conjunction with -c to have custom
startup routines with arguments.
.TP
.B \-h
Show summary of all options.
.TP
.B \-i
Enables unit info dumping into log (for debugging).
.TP
.B \-I addr
With this option you can specify the network address for network play.
.TP
.B \-l
Disable command log messages.
.TP
.B \-N name
The name of the player (will appear in network games).
.TP
.B \-o
Do not use OpenGL or OpenGL ES 1.1.
.TP
.B \-O
Use OpenGL or OpenGL ES 1.1.
.TP
.B \-p
Enables debug messages.
.TP
.B \-P port
With this option you can set the network port to use in network games.
The default value is port udp/6660.
.TP
.B \-s number
Number of frames before AI awakes. This is useful if the AI is too strong and
you want it to react slower. The higher, the easier will be the game against the
AI. Default is to sleep 0 time frames.
.TP
.B \-S number
Sync speed (100 = 30 frames/s).
.TP
.B \-v mode
With this option you can specify video mode resolution in format \fB<xres>\fRx\fB<yres>\fR
.TP
.B \-W
With this option you can start Stratagus in windowed video mode.
.TP
.B \-x filter-idx
In OpenGL mode, this controls the pixel scaling algorithm that is run on the
entire display. This runs on the GPU Fragment shader, so your card and driver
need to support at least OpenGL 1.3.
  -1 - OpenGL native texture scaling (disables scaling in the shader)
  1  - nearest neighbour
  2  - EPX/AdvMame
  3  - HQx
  4  - SAL
  5  - SuperEagle

This is especially useful in combination with -Z, since that will keep the
original size of all elements on the screen at the ratio they have at 640x480,
but scales everything to the chosen video resolution using the nice pixel
scaler. To try which scaler looks best at run-time, you can use Ctrl+Alt+/ to
cycle between scalers.
.TP
.B \-Z
Use OpenGL to scale the screen to the viewport. This is nice for games with
assets only in one resolution, where higher screen resolutions shouldn't make
everything smaller, but instead everything should be scaled up. Implies -O. Use
with -x to get nicer pixel scaling.
.SH "FILES"
.TP
.B ./
Default directory for Stratagus data files
.TP
.B ./scripts/stratagus.lua
Default start config file.
.TP
.B ./scripts/editor.lua
Default editor file.
.TP
.B ./scripts/guichan.lua
Default menu file.
.TP
.B ~/.stratagus/<data>/preferences.lua
User preferences configuration file for <data>
.TP
.B ~/.stratagus/<data>/logs/
User directory with replays/logs for <data>
.TP
.B ~/.stratagus/<data>/save/
User directory with saved games for <data>
.SH "SEE ALSO"
More complete documentation is available in /usr/share/doc/stratagus.
.SH NOTES
Stratagus needs a lot of data (maps and artwork), that can come from
external datasets, like Battle of Survival (BoS), or even from the media
in the Warcraft II CD, thanks to the Wargus dataset.
.PP
Warcraft II is a registered trademark of Blizzard Entertainment.
.SH AUTHOR
Stratagus is written and developed by the Stratagus Team and Contributors.
.PP
This manual page was written by Anthony Towns <ajt@debian.org>,
David Martínez Moreno <ender@debian.org>,
Pali Rohár <pali.rohar@gmail.com>, and
Tim Felgentreff <timfelgentreff@gmail.com> for the Debian GNU/Linux
system (but may be used by others).


CHANGELOG
Change log for Bos Wars
-----------------------

- Fix save games with too many units. bug 35623, patch 8125 (James McCulloch)
- Patch #8319: Fast-forward Through Command Line Option (Douglas Krofcheck)
- Fixed command line speed option (Francois Beerten)
- Added the Cornered map (Iwan Gabovitch)
- Fixed bug #37854: Objectives screen broken: missing text and 
  freeze (Francois Beerten)
- Fixed bug #37861: "Set new units target" makes engineers harvest 
  buildings (Francois Beerten)
- Added the Ringway map (Michiel van der Wulp)
- Added the Wormroads map (Iwan Gabovitch)

Version 2.7 (2013 06 16):
- Added the Boxland map (Iwan Gabovitch)
- Added the islands campaign (Connor MacKay)
- Extended the swindler campaign (Michiel van der Wulp)
- Added sounds for aircraftfactory, artil, destroyer and shipyard (Iwan Gabovitch)
- Do not show waypoints for enemy units when SHIFT key is pressed down (Pali Rohar)
- Added support for 1024x600 and 1366x768 video modes (Kalle Olavi Niemitalo)
- Updated the development documentation (Kalle Olavi Niemitalo)
- Added maps: zombies, cannons, smalland, the small ring (Michiel van der Wulp)
- Improved the clover map (Michiel van der Wulp)
- Added 5 maps from the overpopulation series (Michiel van der Wulp)
- Added goshgeom patches (Michiel van der Wulp)
- Added Lithuanian translation (Vytautas Raškevičius)
- Fixed bug "uncontrollable with Wacom tablet when in full-screen mode"
  as reported at http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=559615 
  (Kalle Olavi Niemitalo)
- Added Swamp Dwellers map (Michiel van der Wulp)
- Added Hungarian translation (Ur Balazs)
- Added the trenches map (Michiel van der Wulp)
- Partially fixed bug #30994: Make AI work with naval units (Kalle Olavi Niemitalo)
- Fixed bug #30971: Units attack buildings too stubbornly (Kalle Olavi Niemitalo)
- Fixed bug #30856: Workers swich to other resourse type after one deposit complete (Kalle Olavi Niemitalo)
- New arrow and cross cursor graphics (Leonid Gutmann)
- Added "terrain" build restriction (Kalle Olavi Niemitalo)
- Added the Tokelau map (Michiel van der Wulp)
- Added the Volcanic Grounds map (Michiel van der Wulp)
- Added the Chilka map (Michiel van der Wulp)
- Added undo groups to the map editor (Kalle Olavi Niemitalo)
- Added ctrl-shift-t hotkey in the editor to tile the selected patch (Jimmy Salmon)
- Added patch themes (Jimmy Salmon)
- Fixed bug #30173: Map editor should require at least one player type to be a person (Jimmy Salmon)
- Fixed bug #29740: Map editor adds units even if dragging began in the sidebar (Jimmy Salmon)
- Fixed bug #30164: Map editor doesn't close terrain file (Jimmy Salmon)
- Fixed bug #30244: Crash with some map sizes in minimap (Francois Beerten)
- Added Clover map (Michiel van der Wulp)
- Fixed bug #30112: How to set the default for "Fog of War" in "Start Game" is not obvious (Kalle Olavi Niemitalo)
- Added AllowTerrain* flags in DefineUnitType (Kalle Olavi Niemitalo)
- Distinguish between shallow and deep water (Kalle Olavi Niemitalo)
- Added Gosh patches (Tejay Penfold)
- Added Shipyard (Tejay Penfold and Francois Beerten)
- Added "Show Terrain Attributes" to the map editor (Kalle Olavi Niemitalo)
- Fixed bug #29731: Replacing patches in the map editor discards CMapField::Flags set because of units (Kalle Olavi Niemitalo)
- Added Multiple Fronts map (Michiel van der Wulp)
- Added a file browser and a size suggestion to Create New Patch (Kalle Olavi Niemitalo)
- Added Atoll Walk map (Michiel van der Wulp)
- Added destroyer ship (Tejay Penfold and Francois Beerten)
- Added water scout unit (Tejay Penfold and Francois Beerten)
- Fixed TextField crash on shift+arrow after backspace (Kalle Olavi Niemitalo)
- Fixed bug #28951: Undo/redo in the editor now supports patches (Jimmy Salmon)
- Added unit-rocksfield to the map editor (Kalle Olavi Niemitalo)
- Updated tolua++ to 1.0.93 (Jimmy Salmon)
- Fixed bug #22211: Arrow keys don't work right when selecting maps (Jimmy Salmon)
- Fixed bug #29096: Adjacent air units flicker in z-order (Jimmy Salmon)
- Fixed bug #29533: No way to exit the campaign in the briefing screen (Jimmy Salmon)
- Fixed bug #29389: Save/recall map position not preserved during save/reload of game (Jimmy Salmon)
- Fixed bug #29284: Custom patches aren't completely cleaned (Jimmy Salmon)
- Build static binary without requiring apg++ from autopackage (Francois Beerten)
- Fixed bug #29100: Cannot translate resource names shown in map editor (Jimmy Salmon)
- Add build script using fabricate (Francois Beerten)
- Add moon patches (Tejay Penfold)
- Fixed bug #29274: Disabling command log crashes and doesn't always work as expected (Jimmy Salmon)
- Fixed bug #29579: KillUnitAt takes 5 parameters (Jimmy Salmon)

Version 2.6.1 (2010 04 18):
- Fix the position of the morels (Kalle Olavi Niemitalo)
- Correct the draw level of some unit types (Kalle Olavi Niemitalo)
- Update the documentation of the scripting API (Francois Beerten)
- On Windows, use the application directory to save user data (Nehal Mistry)

src/stratagus/stratagus.cpp
//       _________ __                 __
//      /   _____//  |_____________ _/  |______     ____  __ __  ______
//      \_____  \\   __\_  __ \__  \\   __\__  \   / ___\|  |  \/  ___/
//      /        \|  |  |  | \// __ \|  |  / __ \_/ /_/  >  |  /\___ |
//     /_______  /|__|  |__|  (____  /__| (____  /\___  /|____//____  >
//             \/                  \/          \//_____/            \/
//  ______________________                           ______________________
//                        T H E   W A R   B E G I N S
//         Stratagus - A free fantasy real time strategy game engine
//
/**@name stratagus.cpp - The main file. */
//
//      (c) Copyright 1998-2015 by Lutz Sammer, Francois Beerten,
//                                 Jimmy Salmon, Pali Rohár and cybermind
//
//      This program is free software; you can redistribute it and/or modify
//      it under the terms of the GNU General Public License as published by
//      the Free Software Foundation; only version 2 of the License.
//
//      This program is distributed in the hope that it will be useful,
//      but WITHOUT ANY WARRANTY; without even the implied warranty of
//      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//      GNU General Public License for more details.
//
//      You should have received a copy of the GNU General Public License
//      along with this program; if not, write to the Free Software
//      Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
//      02111-1307, USA.
//

//@{

/**
** @mainpage
**
** @section Introduction Introduction
**
** Welcome to the source code documentation of the Stratagus engine.
** Extract the source documentation with doxygen (http://www.doxygen.org) tool.
**
** Any help to improve this documention is welcome. If you didn't
** understand something or you found an error or a wrong spelling
** or wrong grammar please write an email (including a patch :).
**
** @section Information Information
**
** Visit the https://github.com/Wargus/stratagus web page for the latest news and
** <A HREF="../index.html">Stratagus Info</A> for other documentation.
**
** @section Modules Modules
**
** This are the main modules of the Stratagus engine.
**
** @subsection Map Map
**
** Handles the map. A map is made from tiles.
**
** @see map.h @see map.cpp @see tileset.h @see tileset.cpp
**
** @subsection Unit Unit
**
** Handles units. Units are ships, flyers, buildings, creatures,
** machines.
**
** @see unit.h @see unit.cpp @see unittype.h @see unittype.cpp
**
** @subsection Missile Missile
**
** Handles missiles. Missiles are all other sprites on map
** which are no unit.
**
** @see missile.h @see missile.cpp
**
** @subsection Player Player
**
** Handles players, all units are owned by a player. A player
** could be controlled by a human or a computer.
**
** @see player.h @see player.cpp @see ::CPlayer
**
** @subsection Sound Sound
**
** Handles the high and low level of the sound. There are the
** background music support, voices and sound effects.
** Following low level backends are supported: OSS and SDL.
**
** @todo adpcm file format support for sound effects
** @todo better separation of low and high level, assembler mixing
** support.
** @todo Streaming support of ogg/mp3 files.
**
** @see sound.h @see sound.cpp
** @see script_sound.cpp @see sound_id.cpp @see sound_server.cpp
** @see unitsound.cpp
** @see sdl_audio.cpp
** @see ogg.cpp @see wav.cpp
**
** @subsection Video Video
**
** Handles the high and low level of the graphics.
** This also contains the sprite and linedrawing routines.
**
** See page @ref VideoModule for more information upon supported
** features and video platforms.
**
** @see video.h @see video.cpp
**
** @subsection Network Network
**
** Handles the high and low level of the network protocol.
** The network protocol is needed for multiplayer games.
**
** See page @ref NetworkModule for more information upon supported
** features and API.
**
** @see network.h @see network.cpp
**
** @subsection Pathfinder Pathfinder
**
** @see pathfinder.h @see pathfinder.cpp
**
** @subsection AI AI
**
** There are currently two AI's. The old one is very hardcoded,
** but does things like placing buildings better than the new.
** The old AI shouldn't be used.  The new is very flexible, but
** very basic. It includes none optimations.
**
** See page @ref AiModule for more information upon supported
** features and API.
**
** @see ai_local.h
** @see ai.h @see ai.cpp
**
** @subsection CCL CCL
**
** CCL is Craft Configuration Language, which is used to
** configure and customize Stratagus.
**
** @see script.h @see script.cpp
**
** @subsection Icon Icon
**
** @see icons.h @see icons.cpp
**
** @subsection Editor Editor
**
** This is the integrated editor, it shouldn't be a perfect
** editor. It is used to test new features of the engine.
**
** See page @ref EditorModule for more information upon supported
** features and API.
**
** @see editor.h @see editor.cpp
*/

/*----------------------------------------------------------------------------
--  Includes
----------------------------------------------------------------------------*/

#include <ctype.h>

#ifdef USE_BEOS
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>

extern void beos_init(int argc, char **argv);

#endif

#ifdef MAC_BUNDLE
#define Button ButtonOSX
#include <Carbon/Carbon.h>
#undef Button
#endif

#include "SDL.h"

#include "stratagus.h"

#include "ai.h"
#include "editor.h"
#include "game.h"
#include "guichan.h"
#include "interface.h"
#include "iocompat.h"
#include "iolib.h"
#include "map.h"
#include "netconnect.h"
#include "network.h"
#include "parameters.h"
#include "player.h"
#include "replay.h"
#include "results.h"
#include "settings.h"
#include "sound_server.h"
#include "title.h"
#include "translate.h"
#include "ui.h"
#include "unit_manager.h"
#include "version.h"
#include "video.h"
#include "widgets.h"
#include "util.h"

#ifdef DEBUG
#include "missile.h" //for FreeBurningBuildingFrames
#endif

#ifdef USE_STACKTRACE
#include <stdexcept>
#include <stacktrace/call_stack.hpp>
#include <stacktrace/stack_exception.hpp>
#endif

#include <stdlib.h>
#include <stdio.h>

#ifdef USE_WIN32
#include <windows.h>
#include <dbghelp.h>
#endif

#if defined(USE_WIN32) && ! defined(NO_STDIO_REDIRECT)
#include "windows.h"
#define REDIRECT_OUTPUT
#endif

#if defined(USE_WIN32) && ! defined(REDIRECT_OUTPUT)
#include "SetupConsole_win32.h"
#endif

/*----------------------------------------------------------------------------
--  Variables
----------------------------------------------------------------------------*/

std::string StratagusLibPath;        /// Path for data directory

/// Name, Version, Copyright
const char NameLine[] = NAME " v" VERSION ", " COPYRIGHT;

std::string CliMapName;          /// Filename of the map given on the command line
std::string MenuRace;

bool EnableDebugPrint;           /// if enabled, print the debug messages
bool EnableAssert;               /// if enabled, halt on assertion failures
bool EnableUnitDebug;            /// if enabled, a unit info dump will be created

/*============================================================================
==  MAIN
============================================================================*/

/**
**  Pre menu setup.
*/
void PreMenuSetup()
{
	//
	//  Initial menus require some gfx.
	//
	SetDefaultTextColors(FontYellow, FontWhite);

	LoadFonts();

	InitVideoCursors();

	if (MenuRace.empty()) {
		LoadCursors(PlayerRaces.Name[0]);
	} else {
		LoadCursors(MenuRace);
	}

	InitSettings();

	InitUserInterface();
	UI.Load();
}

/**
**  Run the guichan main menus loop.
**
**  @return          0 for success, else exit.
*/
static int MenuLoop()
{
	int status;

	initGuichan();
	InterfaceState = IfaceStateMenu;
	//  Clear screen
	Video.ClearScreen();
	Invalidate();

	ButtonUnderCursor = -1;
	OldButtonUnderCursor = -1;
	CursorState = CursorStatePoint;
	GameCursor = UI.Point.Cursor;

	// FIXME delete this when switching to full guichan GUI
	const std::string filename = LibraryFileName("scripts/guichan.lua");
	status = LuaLoadFile(filename);

	// We clean up later in Exit
	return status;
}

//----------------------------------------------------------------------------

/**
**  Print headerline, copyright, ...
*/
static void PrintHeader()
{
	std::string CompileOptions =
#ifdef DEBUG
		"DEBUG "
#endif
#ifdef USE_ZLIB
		"ZLIB "
#endif
#ifdef USE_BZ2LIB
		"BZ2LIB "
#endif
#ifdef USE_VORBIS
		"VORBIS "
#endif
#ifdef USE_THEORA
		"THEORA "
#endif
#ifdef USE_FLUIDSYNTH
		"FLUIDSYNTH "
#endif
#ifdef USE_MIKMOD
		"MIKMOD "
#endif
#ifdef USE_MNG
		"MNG "
#endif
#ifdef USE_OPENGL
		"OPENGL "
#endif
#ifdef USE_GLES
		"GLES "
#endif
#ifdef USE_WIN32
		"WIN32 "
#endif
#ifdef USE_LINUX
		"LINUX "
#endif
#ifdef USE_BSD
		"BSD "
#endif
#ifdef USE_BEOS
		"BEOS "
#endif
#ifdef USE_MAC
		"MAC "
#endif
#ifdef USE_X11
		"X11 "
#endif
#ifdef USE_TOUCHSCREEN
		"TOUCHSCREEN "
#endif
		"";

	fprintf(stdout,
			"%s\n  written by Lutz Sammer, Fabrice Rossi, Vladi Shabanski, Patrice Fortier,\n"
			"  Jon Gabrielson, Andreas Arens, Nehal Mistry, Jimmy Salmon, Pali Rohar,\n"
			"  cybermind and others.\n"
			"\t" HOMEPAGE "\n"
			"Compile options %s",
			NameLine, CompileOptions.c_str());
}

void PrintLicense()
{
	printf("\n"
		   "\n"
		   "Stratagus may be copied only under the terms of the GNU General Public License\n"
		   "which may be found in the Stratagus source kit.\n"
		   "\n"
		   "DISCLAIMER:\n"
		   "This software is provided as-is.  The author(s) can not be held liable for any\n"
		   "damage that might arise from the use of this software.\n"
		   "Use it at your own risk.\n"
		   "\n");
}


/**
**  Exit the game.
**
**  @param err  Error code to pass to shell.
*/
void Exit(int err)
{
	if (GameRunning) {
		StopGame(GameExit);
		return;
	}

	StopMusic();
	QuitSound();
	NetworkQuitGame();

	ExitNetwork1();
#ifdef DEBUG
	CleanModules();
	FreeBurningBuildingFrames();
	FreeSounds();
	FreeGraphics();
	FreePlayerColors();
	FreeButtonStyles();
	FreeAllContainers();
	freeGuichan();
	DebugPrint("Frames %lu, Slow frames %d = %ld%%\n" _C_
			   FrameCounter _C_ SlowFrameCounter _C_
			   (SlowFrameCounter * 100) / (FrameCounter ? FrameCounter : 1));
	lua_settop(Lua, 0);
	lua_close(Lua);
	DeInitVideo();
#endif

	fprintf(stdout, "%s", _("Thanks for playing Stratagus.\n"));
	exit(err);
}

/**
**  Do a fatal exit.
**  Called on out of memory or crash.
**
**  @param err  Error code to pass to shell.
*/
void ExitFatal(int err)
{
#ifdef USE_STACKTRACE
	throw stacktrace::stack_runtime_error((const char*)err);
#endif
	exit(err);
}

/**
**  Display the usage.
*/
static void Usage()
{
	PrintHeader();
	printf(
		"\n\nUsage: %s [OPTIONS] [map.smp|map.smp.gz]\n"
		"\t-a\t\tEnables asserts check in engine code (for debugging)\n"
		"\t-c file.lua\tConfiguration start file (default stratagus.lua)\n"
		"\t-d datapath\tPath to stratagus data (default current directory)\n"
		"\t-D depth\tVideo mode depth = pixel per point\n"
		"\t-e\t\tStart editor (instead of game)\n"
		"\t-E file.lua\tEditor configuration start file (default editor.lua)\n"
		"\t-F\t\tFull screen video mode\n"
		"\t-G \"options\"\tGame options (passed to game scripts)\n"
		"\t-h\t\tHelp shows this page\n"
		"\t-i\t\tEnables unit info dumping into log (for debugging)\n"
		"\t-I addr\t\tNetwork address to use\n"
		"\t-l\t\tDisable command log\n"
		"\t-N name\t\tName of the player\n"
#if defined(USE_OPENGL) || defined(USE_GLES)
		"\t-o\t\tDo not use OpenGL or OpenGL ES 1.1\n"
		"\t-O\t\tUse OpenGL or OpenGL ES 1.1\n"
#endif
		"\t-p\t\tEnables debug messages printing in console\n"
		"\t-P port\t\tNetwork port to use\n"
		"\t-s sleep\tNumber of frames for the AI to sleep before it starts\n"
		"\t-S speed\tSync speed (100 = 30 frames/s)\n"
		"\t-u userpath\tPath where stratagus saves preferences, log and savegame\n"
		"\t-v mode\t\tVideo mode resolution in format <xres>x<yres>\n"
		"\t-W\t\tWindowed video mode\n"
#if defined(USE_OPENGL) || defined(USE_GLES)
		"\t-x idx\t\tControls fullscreen scaling if your graphics card supports shaders.\n"\
		"\t  \t\tPass 1 for nearest-neigubour, 2 for EPX/AdvMame, 3 for HQx, 4 for SAL, 5 for SuperEagle\n"\
		"\t  \t\tYou can also use Ctrl+Alt+/ to cycle between these scaling algorithms at runtime.\n"
		"\t  \t\tPass -1 to force old-school nearest neighbour scaling without shaders\n"\
		"\t-Z\t\tUse OpenGL to scale the screen to the viewport (retro-style). Implies -O.\n"
#endif
		"map is relative to StratagusLibPath=datapath, use ./map for relative to cwd\n",
		Parameters::Instance.applicationName.c_str());
}

#ifdef REDIRECT_OUTPUT

static std::string stdoutFile;
static std::string stderrFile;

static void CleanupOutput()
{
	fclose(stdout);
	fclose(stderr);

	struct stat st;
	if (stat(stdoutFile.c_str(), &st) == 0 && st.st_size == 0) {
		unlink(stdoutFile.c_str());
	}
	if (stat(stderrFile.c_str(), &st) == 0 && st.st_size == 0) {
		unlink(stderrFile.c_str());
	}
}

static void RedirectOutput()
{
	char path[MAX_PATH];
	int pathlen;

	pathlen = GetModuleFileName(NULL, path, sizeof(path));
	while (pathlen > 0 && path[pathlen] != '\\') {
		--pathlen;
	}
	path[pathlen] = '\0';

	stdoutFile = std::string(path) + "\\stdout.txt";
	stderrFile = std::string(path) + "\\stderr.txt";

	if (!freopen(stdoutFile.c_str(), "w", stdout)) {
		printf("freopen stdout failed");
	}
	if (!freopen(stderrFile.c_str(), "w", stderr)) {
		printf("freopen stderr failed");
	}
	atexit(CleanupOutput);
}
#endif

void ParseCommandLine(int argc, char **argv, Parameters &parameters)
{
	for (;;) {
		switch (getopt(argc, argv, "ac:d:D:eE:FG:hiI:lN:oOP:ps:S:u:v:Wx:Z?-")) {
			case 'a':
				EnableAssert = true;
				continue;
			case 'c':
				parameters.luaStartFilename = optarg;
				if (strlen(optarg) > 4 &&
				    !(strstr(optarg, ".lua") == optarg + strlen(optarg) - 4)) {
					parameters.luaStartFilename += ".lua";
				}
				continue;
			case 'd': {
				StratagusLibPath = optarg;
				size_t index;
				while ((index = StratagusLibPath.find('\\')) != std::string::npos) {
					StratagusLibPath[index] = '/';
				}
				continue;
			}
			case 'D':
				Video.Depth = atoi(optarg);
				continue;
			case 'e':
				Editor.Running = EditorCommandLine;
				continue;
			case 'E':
				parameters.luaEditorStartFilename = optarg;
				continue;
			case 'F':
				VideoForceFullScreen = 1;
				Video.FullScreen = 1;
				continue;
			case 'G':
				parameters.luaScriptArguments = optarg;
				continue;
			case 'i':
				EnableUnitDebug = true;
				continue;
			case 'I':
				CNetworkParameter::Instance.localHost = optarg;
				continue;
			case 'l':
				CommandLogDisabled = true;
				continue;
			case 'N':
				parameters.LocalPlayerName = optarg;
				continue;
#if defined(USE_OPENGL) || defined(USE_GLES)
			case 'o':
				ForceUseOpenGL = 1;
				UseOpenGL = 0;
				if (ZoomNoResize) {
					fprintf(stderr, "Error: -Z only works with OpenGL enabled\n");
					Usage();
					ExitFatal(-1);
				}
				continue;
			case 'O':
				ForceUseOpenGL = 1;
				UseOpenGL = 1;
				continue;
#endif
			case 'P':
				CNetworkParameter::Instance.localPort = atoi(optarg);
				continue;
			case 'p':
				EnableDebugPrint = true;
				continue;
			case 's':
				AiSleepCycles = atoi(optarg);
				continue;
			case 'S':
				VideoSyncSpeed = atoi(optarg);
				continue;
			case 'u':
				Parameters::Instance.SetUserDirectory(optarg);
				continue;
			case 'v': {
				char *sep = strchr(optarg, 'x');
				if (!sep || !*(sep + 1)) {
					fprintf(stderr, "%s: incorrect format of video mode resolution -- '%s'\n", argv[0], optarg);
					Usage();
					ExitFatal(-1);
				}
				Video.Height = atoi(sep + 1);
				*sep = 0;
				Video.Width = atoi(optarg);
				if (!Video.Height || !Video.Width) {
					fprintf(stderr, "%s: incorrect format of video mode resolution -- '%sx%s'\n", argv[0], optarg, sep + 1);
					Usage();
					ExitFatal(-1);
				}
#if defined(USE_OPENGL) || defined(USE_GLES)
				if (ZoomNoResize) {
					Video.ViewportHeight = Video.Height;
					Video.ViewportWidth = Video.Width;
					Video.Height = 0;
					Video.Width = 0;
				}
#endif
				continue;
			}
			case 'W':
				VideoForceFullScreen = 1;
				Video.FullScreen = 0;
				continue;
#if defined(USE_OPENGL) || defined(USE_GLES)
			case 'x':
				ShaderIndex = atoi(optarg) % MAX_SHADERS;
				if (atoi(optarg) == -1) {
					GLShaderPipelineSupported = false;
				}
				continue;
			case 'Z':
				ForceUseOpenGL = 1;
				UseOpenGL = 1;
				ZoomNoResize = 1;
				Video.ViewportHeight = Video.Height;
				Video.ViewportWidth = Video.Width;
				Video.Height = 0;
				Video.Width = 0;
				continue;
#endif
			case -1:
				break;
			case '?':
			case 'h':
			default:
				Usage();
				ExitFatal(-1);
		}
		break;
	}

	if (argc - optind > 1) {
		fprintf(stderr, "too many map files. if you meant to pass game arguments, these go after '--'\n");
		Usage();
		ExitFatal(-1);
	}

	if (argc - optind) {
		size_t index;
		CliMapName = argv[optind];
		while ((index = CliMapName.find('\\')) != std::string::npos) {
			CliMapName[index] = '/';
		}
	}
}

#ifdef USE_WIN32
static LONG WINAPI CreateDumpFile(EXCEPTION_POINTERS *ExceptionInfo)
{
	HANDLE hFile = CreateFile("crash.dmp", GENERIC_READ | GENERIC_WRITE,	FILE_SHARE_DELETE | FILE_SHARE_READ | FILE_SHARE_WRITE,
		NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL,	NULL);
	MINIDUMP_EXCEPTION_INFORMATION mei;
	mei.ThreadId = GetCurrentThreadId();
	mei.ClientPointers = TRUE;
	mei.ExceptionPointers = ExceptionInfo;
	MiniDumpWriteDump(GetCurrentProcess(), GetCurrentProcessId(), hFile, MiniDumpNormal, &mei, NULL, NULL);
	fprintf(stderr, "Stratagus crashed!\n");
	fprintf(stderr, "A mini dump file \"crash.dmp\" has been created in the Stratagus folder.\n");
	fprintf(stderr, "Please send it to our bug tracker: https://github.com/Wargus/stratagus/issues\n");
	fprintf(stderr, "and tell us what caused this bug to occur.\n");
	return EXCEPTION_EXECUTE_HANDLER;
}
#endif

/**
**  The main program: initialise, parse options and arguments.
**
**  @param argc  Number of arguments.
**  @param argv  Vector of arguments.
*/
int stratagusMain(int argc, char **argv)
{
#ifdef REDIRECT_OUTPUT
	RedirectOutput();
#endif
#ifdef USE_BEOS
	//  Parse arguments for BeOS
	beos_init(argc, argv);
#endif
#ifdef USE_WIN32
	SetUnhandledExceptionFilter(CreateDumpFile);
#endif
#if defined(USE_WIN32) && ! defined(REDIRECT_OUTPUT)
	SetupConsole();
#endif
	//  Setup some defaults.
#ifndef MAC_BUNDLE
	StratagusLibPath = ".";
#else
	freopen("/tmp/stdout.txt", "w", stdout);
	freopen("/tmp/stderr.txt", "w", stderr);
	// Look for the specified data set inside the application bundle
	// This should be a subdir of the Resources directory
	CFURLRef pluginRef = CFBundleCopyResourceURL(CFBundleGetMainBundle(),
												 CFSTR(MAC_BUNDLE_DATADIR), NULL, NULL);
	CFStringRef macPath = CFURLCopyFileSystemPath(pluginRef,  kCFURLPOSIXPathStyle);
	const char *pathPtr = CFStringGetCStringPtr(macPath, CFStringGetSystemEncoding());
	Assert(pathPtr);
	StratagusLibPath = pathPtr;
#endif
#ifdef USE_STACKTRACE
	try {
#endif
	Parameters &parameters = Parameters::Instance;
	parameters.SetDefaultValues();
	parameters.SetLocalPlayerNameFromEnv();

	if (argc > 0) {
		parameters.applicationName = argv[0];
	}

	// FIXME: Parse options before or after scripts?
	ParseCommandLine(argc, argv, parameters);
	// Init the random number generator.
	InitSyncRand();

	makedir(parameters.GetUserDirectory().c_str(), 0777);

	// Init Lua and register lua functions!
	InitLua();
	LuaRegisterModules();

	// Initialise AI module
	InitAiModule();

	LoadCcl(parameters.luaStartFilename, parameters.luaScriptArguments);

	PrintHeader();
	PrintLicense();

	// Setup video display
	InitVideo();

	// Setup sound card
	if (!InitSound()) {
		InitMusic();
	}

#ifndef DEBUG           // For debug it's better not to have:
	srand(time(NULL));  // Random counter = random each start
#endif

	//  Show title screens.
	SetDefaultTextColors(FontYellow, FontWhite);
	LoadFonts();
	SetClipping(0, 0, Video.Width - 1, Video.Height - 1);
	Video.ClearScreen();
	ShowTitleScreens();

	// Init player data
	ThisPlayer = NULL;
	//Don't clear the Players structure as it would erase the allowed units.
	// memset(Players, 0, sizeof(Players));
	NumPlayers = 0;

	UnitManager.Init(); // Units memory management
	PreMenuSetup();     // Load everything needed for menus

	MenuLoop();

	Exit(0);
#ifdef USE_STACKTRACE
	} catch (const std::exception &e) {
		fprintf(stderr, "Stratagus crashed!\n");
		fprintf(stderr, "Please send this call stack to our bug tracker: https://github.com/Wargus/stratagus/issues\n");
		fprintf(stderr, "and tell us what caused this bug to occur.\n");
		fprintf(stderr, " === exception state traceback === \n");
		fprintf(stderr, "%s", e.what());
		exit(1);
	}
#endif
	return 0;
}

//@}
